docs/proposals/20191017-kubeadm-based-control-plane.md

```diff
@@ -173,12 +178,45 @@ With the following validations:
 - `KubeadmControlPlane.Spec.KubeadmConfigSpec` allows mutations required for supporting following use cases:
     - Change of imagesRepository/imageTags (with validation of CoreDNS supported upgrade)
     - Change of node registration options
-    - Change of pre/post kubeadm commands 
+    - Change of pre/post kubeadm commands
     - Change of cloud init files
 
 And the following defaulting:
 
-- `KubeadmControlPlane.Spec.Replicas: 1` 
+- `KubeadmControlPlane.Spec.Replicas: 1`
+
+##### Rollout strategies
+
+```go
+type RolloutType string
+
+const (
+
+  // CreateFirstRolloutType constant is used to modifying the behavior of control plane during rollout.  
+  // CreateFirst logic first creates a new control plane machine and then deletes one of the old ones.
+  // This is also the default rollout strategy.
+  CreateFirstRolloutType  RolloutType = "CreateFirst"
```

Comment:
I'm wondering if instead of `CreateFirst` and `DeleteFirst` it might be better to call these something like `PreferSurge`/`PreferConstrict` or even take advantage of the api change to v1alpha4 to switch to something like `MaxSurge` and `MaxUnavailable` to align closer with Deployments for this purpose?

I think `PreferSurge`/`PreferConstrict` would more align with the current implementation details around rollout being purely scaleUp/scaleDown operations with manipulated members, which depending on certain edge case scenarios might not actually be the operations being performed (for example if scaling operations are in flight when a rollout is triggered)

If we go with `MaxSurge`/`MaxUnavailable` we could add defaults/validation safeguards, or even a special `preserveQuorum` default value.  This would also allow us to more accurately model the rollout process compared to the existing implementation, allow us to potentially surge/constrict rollouts more flexibly for larger scale deployments (potentially allowing for future parallelization),  while also allowing us to eventually potentially leverage MachineDeployments as an underlying implementation detail for KCP. It would require more invasive changes to the controller logic, but I think it might provide more benefits and flexibility in the long term.

Reply:
Similar discussion above which I started the other day https://github.com/kubernetes-sigs/cluster-api/pull/3857#issuecomment-735620261, ultimately there was enough consensus to prefer strings for now

Reply:
I don't have anything against Deployment like mechanism with MaxSurge/MaxUnavailable fields. Original idea of this proposal was to enable scale-in to make life easier in resource constrained enviroments. I have to admit this might be little short-sighted, but much needed thou.

Now that this Proposal has drawn more attention, I agree we should discuss long term goals. If we found Deployment like behaviour more usefull in KCP, we should defenitely go with it.

My only concern is how we prevent users to missconfigure the rollout with MaxSurge/MaxUnavailable? But like Jason said maybe we could add defaults/validation safeguards for preventing missconfigurations.

Reply:
> My only concern is how we prevent users to misconfigure the rollouts with MaxSurge/MaxUnavailable? But like @detiber said maybe we could add defaults/validation safeguards for preventing misconfigurations.

We can probably leverage defaulting + validation for these, and reject any updates that would potentially be harmful for the system